editor/src/messages/portfolio/document/node_graph/node_properties.rs

```diff
@@ -1903,44 +1833,37 @@ pub fn math_properties(node_id: NodeId, context: &mut NodePropertiesContext) ->
 }
 
 pub struct ParameterWidgetsInfo<'a> {
-	document_node: &'a DocumentNode,
+	document_node: Option<&'a DocumentNode>,
 	node_id: NodeId,
 	index: usize,
-	name: &'a str,
-	description: &'a str,
+	name: String,
+	description: String,
+	input_type: FrontendGraphDataType,
 	blank_assist: bool,
+	exposeable: bool,
 }
 
 impl<'a> ParameterWidgetsInfo<'a> {
-	pub fn new(document_node: &'a DocumentNode, node_id: NodeId, index: usize, name: &'a str, description: &'a str, blank_assist: bool) -> ParameterWidgetsInfo<'a> {
-		ParameterWidgetsInfo {
-			document_node,
-			node_id,
-			index,
-			name,
-			description,
-			blank_assist,
-		}
-	}
+	pub fn new(node_id: NodeId, index: usize, blank_assist: bool, context: &'a mut NodePropertiesContext) -> ParameterWidgetsInfo<'a> {
```

Comment:
Create a single constructor which merges all the logic which was previously performed outside